### ISSUE TITLE
.onLoad import_from_path works in R 3.6 but not 4.0.1

### ISSUE THREAD

**Issue by johnERgordon** [rstudio/reticulate#780]
Hi, 
When I use R 3.6 I can build a package with below script (that I got from here: https://community.rstudio.com/t/build-package-namespace-dynamically-during-onload/4101/3 Note I've also tried the other alternatives on this page...) 
 
**But I get the following error in R 4.01.** 
** byte-compile and prepare package for lazy loading
sh: line 1: 43728 Killed: 9               R_TESTS= '/Library/Frameworks/R.framework/Resources/bin/R' --no-save --no-restore --no-echo 2>&1 < '/var/folders/y7/wj3424xx4y3_4zsn044s7r5h0000gn/T//RtmpWEaHSn/fileaac8c10a8b3'
ERROR: lazy loading failed for package ‘testPackage’

# Load the module and create dummy objects from it, all of which are NULL
the_py_module <- reticulate::import_from_path(
  "the_py_module",
  file.path("inst", "python")
)
for (obj in names(the_py_module)) {
  assign(obj, NULL)
}
# Clean up
rm(the_py_module)

# Now all those names are in the namespace, and ready to be replaced on load
.onLoad <- function(libname, pkgname) {
  the_py_module <- reticulate::import_from_path(
    "the_py_module",
    system.file("python", package = "text")
  )
  # assignInMyNamespace(...) is meant for namespace manipulation
  for (obj in names(the_py_module)) {
    assignInMyNamespace(obj, the_py_module[[obj]])
  }
}

Any help is much appreciated; thanks.

**Comment by johnERgordon** [comment]
I have now tried the below more simplified example:

**In a inst/python/the_py_module.py file:**

```
def f1():
  return "f one"

def f2():
  return "f two"
```

 
**In a R/zzz.R file**
```
f1 <- NULL
the_module <- NULL

.onLoad <- function(libname, pkgname) {
    the_module <- reticulate::import_from_path(
      "the_py_module",
      file.path("inst", "python"))
    f1 <<- the_module$f1
}
```


**In a R/functions.R file** 
```
#  devtools::document()
#' Testing F1 function from python
#'
#' @return "f one"
#' @export
f1_from_python <- function(){
  f1()
  }
# f1_from_python()
```

Building the package (Command+shift+B) runs in R 3.6 and running `f1_from_python()` then returns "f one". 

However, building it in R4.02 give the following error. 

```
h: line 1: 67874 Killed: 9               R_TESTS= '/Library/Frameworks/R.framework/Resources/bin/R' --no-save --no-restore --no-echo 2>&1 < '/var/folders/y7/wj3424xx4y3_4zsn044s7r5h0000gn/T//Rtmpbtc8AH/file1090c542d0309'
ERROR: loading failed
* removing ‘/Library/Frameworks/R.framework/Versions/4.0/Resources/library/test_function’
* restoring previous ‘/Library/Frameworks/R.framework/Versions/4.0/Resources/library/test_function’

Exited with status 1.
```

**Comment by OscarKjell** [comment]
Have there been any update on this?

**Comment by kevinushey** [comment]
Sorry, no update yet. I fear this is probably caused by newer builds of R on macOS being notarized, and I don't know a solution yet.

Does it make a difference if you delay loading the module, as described in https://rstudio.github.io/reticulate/articles/package.html?

**Comment by johnERgordon** [comment]
@kevinushey  thanks for the suggestion; but unfortunately it did not work loading the module with:

```
.onLoad <- function(libname, pkgname) {
  the_module <- reticulate::import(
    "the_py_module", delay_load = TRUE)
  f1 <<- the_module$f1
}
```
Got a similar error: 

** testing if installed package can be loaded from temporary location
sh: line 1: 87630 Killed: 9               R_TESTS= '/Library/Frameworks/R.framework/Resources/bin/R' --no-save --no-restore --no-echo 2>&1 < '/var/folders/y7/wj3424xx4y3_4zsn044s7r5h0000gn/T//RtmpMJNNjt/file1563bbd9cfc3'
ERROR: loading failed

Any other suggestions on solution please let us know. 
Thanks

**Comment by kevinushey** [comment]
The problem is that calling `f1 <<- the_module$f1` basically forces an eager load of the module. You'd want to avoid accessing or using the module in your `.onLoad()` and perform any access to the lazy-loaded module within your package's exported R functions.

**Comment by johnERgordon** [comment]
Thank you for the help! 
This now works: 
```
f1 <- NULL
the_module <- NULL

.onLoad <- function(libname, pkgname) {
  the_module <<- reticulate::import(
    "the_py_module", delay_load = TRUE)
}
```
However:
1. this requires putting  the_py_module.py file in the root folder; but CRAN requires this type of file to be in the inst/python directory. And using import_from_path() does not allow to set the ```delay_load = TRUE``` parameter. Am I missing something here? 

2. before I used reticulate::source_python() within the exported package function. But not using reticulate::source_python() and instead using reticulate::import() gives an error because it is not possible to set an objected needed for the python script. It throws the following error: object 'GeX' not found.

3. What is the difference between reticulate::source_python() and reticulate::import() in this sense; I reckon the first requires running the python script everytime the R-package function is used (which takes some time); but is there any other negative aspects to this?
 
Any guidance is much appreciated. 
Thanks in advance.

**Comment by kevinushey** [comment]
Ahh, I think you're right -- if you need to import a module provided with your package, there isn't a simple way to delay-load that module. I need to think of an appropriate way to make this possible.

The main difference between `source_python()` and `import` is:

1. `source_python()` (ultimately) uses to the Python C API [PyRun_StringFlags](https://docs.python.org/3/c-api/veryhigh.html#c.PyRun_StringFlags) to run the code in the main module;

2. `import()` uses [PyImport_ImportModule](https://docs.python.org/3/c-api/import.html#c.PyImport_ImportModule), which is effectively the same as running `import <module>` in a Python session.

In other words, `import()` uses regular semantics for module encapsulation, whereas `source_python()` just runs Python code in the session.
